Enable rust native `Target` creation.

raynelfss opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary
The following commits expose rust native methods `add_instruction` and `update_instruction_properties`, opposite to their Python counterparts.

### Details and comments
Now that we're able to build instructions efficiently from Rust, we should be able to add instructions to the `Target` (or the python's `BaseTarget`). The following commits add the following Rust-native methods:

---
#### `add_instruction`

Adds an instance of `PackedOperation` to the `Target`. 

An addition to the Target from Python typically involves using an instance of `TargetOperation`, which has two variants representing variadics and regular instructions. However, since we don't yet support `Variadics` in Rust, this method is only able to work with `NormalOperation` and therefore accepts the Rust native types `PackedOperation` and a slice of `Param` (at least until #13267 is resolved).

#### `update_instruction_properties`

Modifies the `InstructionProperties` assigned of a collection of `qargs` for a certain instruction in the `Target`. **This should be the only way of modifying instruction properties**, at least until we can figure out a way of synchronizing the Python and Rust gate maps (see #14173).

---
These additions rely on the additions described below:

### Pre-requisites
- [x] #13995
- [x] #14210

raynelfss committed:
Add: Rust native `add_instruction` to `Target`
- Add rust native way of adding an instruction to the `Target` without any need for Python.
- Additions are guaranteed to be reflected in Rust and Python.

raynelfss committed:
Add: rust native `update_instruction_properties`

raynelfss committed:
Fix: Add Rust native testing for new methods.
- Remove `PyResult` from the output of the rust native functions.
- Add quotation marks to the output of error messages for gate names.
- Add rust native unit testing for the new methods to make sure they're working properly.

mtreinish commented on crates/accelerate/src/target_transpiler/errors.rs:
```diff
@@ -31,4 +31,23 @@ impl Display for TargetKeyError {
-impl Error for TargetKeyError {}
+/// Error thrown when an invalid instruction is added to the [Target]
+#[derive(Debug)]
+pub struct TargetInvalidInstError {
```
We already depend on this error: https://docs.rs/thiserror/latest/thiserror/ you can use it to clean up most of the boilerplate code here.

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -966,6 +898,129 @@ impl Target {
 impl Target {
+    /// Adds a [PackedOperation] to the [Target].
+    ///
+    /// Said addition results in a [NormalOperation] in the [Target] as variadics
```
Can you use a normal rustdoc formatting here for the arguments and return type? This is our internal public interface for adding instructions so having a built more detail and an example would be good.

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -966,6 +898,129 @@ impl Target {
+        Ok(())
+    }
+
+    /// Update the property object for an instruction qarg pair already in the [Target].
```
Same thing here about the docs.

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -966,6 +898,129 @@ impl Target {
+        let name = if let Some(name) = name {
+            name.to_owned()
+        } else {
+            operation.name().to_string()
```
Why `to_string()` here vs `to_owned()` above? Both branches you're converting an `&str` to a `String`.

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -1379,3 +1442,169 @@ pub fn target(m: &Bound<PyModule>) -> PyResult<()> {
 }
 
 // TODO: Add rust-based unit testing.
```
I think you can remove this now.

mtreinish reviewed (commented):
I took a quick first pass through this it's looking good, just some initial comments inline.

raynelfss committed:
Add: Use `this_error` to implement `Target` errors.
- Use rustdoc format for docstrings
- Removce generic type for property map in `add_instruction`.
- Add `QargsRef::as_owned_qargs` to return an owned version of Qargs.

mtreinish commented on crates/accelerate/src/target_transpiler/errors.rs:
```diff
@@ -10,25 +10,35 @@
+    /// An already existing instruction name being queried into the [Target].
+    #[error["Instruction '{0}' is already in the target."]]
+    AlreadyExists(String),
+    /// An attempt to add collection of qargs to the [Target] that do
```
```suggestion
    /// An attempt to add collection of qargs to the [Target] that does
```

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -416,33 +365,15 @@ impl Target {
     ///     KeyError: If ``instruction`` or ``qarg`` are not in the target
-    #[pyo3(signature = (instruction, qargs, properties))]
-    fn update_instruction_properties(
+    #[pyo3(name = "update_instruction_properties", signature = (instruction, qargs, properties))]
+    fn py_update_instruction_properties(
```
Is the only reason your doing this to change the result type? The signature looks the same on both methods so I wanted to make sure we were doing this for a specific reason, instead of just making it `pub`

raynelfss replied:
Yes, not only that but also the fact that we'd preferably use a reference for `Qargs` instead of an owned value which is the only one that can be extracted from Python based on it's `FromPyObject<'py>` implementation.

mtreinish commented on crates/accelerate/src/target_transpiler/qargs.rs:
```diff
@@ -132,6 +132,16 @@ impl QargsRef<'_> {
+
+    /// Converts into an owned version of [Qargs]
+    pub fn as_owned_qargs(&self) -> Qargs {
+        match self {
```
Should this just be the `ToOwned` trait or something similar. The method name here `as_owned_qargs` is a bit weird looking, especially when considering: https://rust-lang.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv it would normally be called `to_owned()`

That being said isn't this just used for error messages, it feels like maybe we should adjust or loosen the typing on the error type arguments so we can just call debug on the `QargsRef` instead of requiring an owned copy.

raynelfss replied:
I agree on us improving the typings on the errors here to avoid the need for this method, since that's the only reason we use it. 

I wanted to have the method named `to_owned` but didn't want to imply that this implements `ToOwned<Owned = Qargs>`mainly because we're unable to implement `Borrow<QargsRef>` for `Qargs`. That said, I'm more than happy to change/remove the method.

raynelfss replied:
Removed in 4ae2e71

mtreinish reviewed (commented):
This is looking good, thanks for making the updates. I just have a few more small comments inline.

raynelfss committed:
Update crates/accelerate/src/target_transpiler/errors.rs

Co-authored-by: Matthew Treinish <mtreinish@kortar.org>

raynelfss committed:
Fix: Address review comments.
- Remove `as_owned_qargs`.
- Use `String` instead of owned `Qargs` for `TargetError`.

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -966,6 +897,179 @@ impl Target {
+    /// * `name` - The name of the instruction if differs from the [PackedOperation]
+    ///   instance.
+    /// * `props_map`: The optional property mapping between [Qargs] and
+    ///   [InstructionProperties].
```
```suggestion
    ///   [InstructionProperties]. If set to `None` the instruction is treated as a global ideal instruction.
```

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -966,6 +897,179 @@ impl Target {
+    /// * `operation` - The [PackedOperation] to be added.
+    /// * `params` - The collection of [Param] assigned to the instruction.
+    /// * `name` - The name of the instruction if differs from the [PackedOperation]
+    ///   instance.
```
```suggestion
    ///   instance. If set to `None` it defaults to the string returned by [`Operation::name`] for `operation`.
```

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -966,6 +897,179 @@ impl Target {
+    /// * `instruction` - The instruction's name within this instance.
+    /// * `qargs` - A collection of [PhysicalQubit] or an instance of [Qargs::Global]
+    ///   that the instruction operated on.
+    ///
```
```suggestion
    /// * `properties` - The properties to use for updating the specified instruction in the target.
    
```

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -966,6 +897,179 @@ impl Target {
+        &mut self,
+        instruction: &'a str,
+        qargs: T,
+        properties: Option<InstructionProperties>,
```
Do we need the `Option` here? Like I realize the data model supports `None` to indicate an ideal instruction with no properties, but I wonder if in practice this is something we should be supporting on an update method. The python space typing indicated `None` wasn't an allowed type only `InstructionProperties`.

raynelfss replied:
That's a good question. Getting rid of the option seems logical. However, what if the user were to update the properties of a global/variadic instruction temporarily, and then back to its original value `None`? What could the user do in this case and is this supported behavior?

mtreinish replied:
I'm fine leaving it. I don't actually think that's a thing people will do in practice. The handling of globals is ill defined especially with properties set and it's not commonly used at all. That was not part of the original design intent of the target, and the fact it was possible was an implementation mistake that leaves us with undefined behavior. This is why I asked you to open #14268 so we can fix this hole in the API for 3.0 (regardless of what we do it will probably be backwards incompatible).

raynelfss committed:
Apply suggestions from code review

Co-authored-by: Matthew Treinish <mtreinish@kortar.org>

raynelfss committed:
Docs: Fix docstring formatting for `update` method

raynelfss committed:
Merge branch 'main' into target-rust-add

mtreinish reviewed (approved):
LGTM, thanks for doing this.
